{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Putative Cargo Loading Peptides\n",
    "\n",
    "We have searched a few different putative cargo loading peptide (CLP) and family 2 N-terminal domain (NTD) motifs against all of our cargo proteins. Let's check out these search hits and see if any of them give clues about potential encapsulin cargo proteins.\n",
    "\n",
    "First, we can load our CLP hits data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>Target</th>\n",
       "      <th>Identity</th>\n",
       "      <th>Alignment Length</th>\n",
       "      <th>Mismatches</th>\n",
       "      <th>Gap Openings</th>\n",
       "      <th>Query Start</th>\n",
       "      <th>Query End</th>\n",
       "      <th>Target Start</th>\n",
       "      <th>Target End</th>\n",
       "      <th>E-Value</th>\n",
       "      <th>Bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP003667212098</td>\n",
       "      <td>0.923</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>114</td>\n",
       "      <td>125</td>\n",
       "      <td>0.001832</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP001555959371</td>\n",
       "      <td>0.846</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>337</td>\n",
       "      <td>348</td>\n",
       "      <td>0.087220</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP003362106646</td>\n",
       "      <td>0.909</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>357</td>\n",
       "      <td>366</td>\n",
       "      <td>0.188500</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP001599236885</td>\n",
       "      <td>0.833</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>336</td>\n",
       "      <td>346</td>\n",
       "      <td>0.407200</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP001627912779</td>\n",
       "      <td>0.875</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>120</td>\n",
       "      <td>127</td>\n",
       "      <td>1.290000</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Query            Target  Identity  Alignment Length  \\\n",
       "0  all_cargo_clp_consensus  MGYP003667212098     0.923                13   \n",
       "1  all_cargo_clp_consensus  MGYP001555959371     0.846                13   \n",
       "2  all_cargo_clp_consensus  MGYP003362106646     0.909                11   \n",
       "3  all_cargo_clp_consensus  MGYP001599236885     0.833                12   \n",
       "4  all_cargo_clp_consensus  MGYP001627912779     0.875                 8   \n",
       "\n",
       "   Mismatches  Gap Openings  Query Start  Query End  Target Start  Target End  \\\n",
       "0           0             1            1         13           114         125   \n",
       "1           1             1            1         13           337         348   \n",
       "2           0             1            1         11           357         366   \n",
       "3           1             1            1         12           336         346   \n",
       "4           1             0            6         13           120         127   \n",
       "\n",
       "    E-Value  Bitscore  \n",
       "0  0.001832        33  \n",
       "1  0.087220        28  \n",
       "2  0.188500        27  \n",
       "3  0.407200        26  \n",
       "4  1.290000        24  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clp_df = pd.read_csv(\"../clp_cargo_hits.m8\", sep=\"\\t\", names=[\"Query\", \"Target\", \"Identity\", \"Alignment Length\", \"Mismatches\", \"Gap Openings\",\n",
    "                                                              \"Query Start\", \"Query End\", \"Target Start\", \"Target End\", \"E-Value\", \"Bitscore\"])\n",
    "\n",
    "clp_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's add in the length of each target protein so we can find out how close our hits are to the C-terminus of the protein.\n",
    "\n",
    "(Note that we don't need to calculate the distance from the N-terminus since the `Target Start` field already contains this information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>Target</th>\n",
       "      <th>Identity</th>\n",
       "      <th>Alignment Length</th>\n",
       "      <th>Mismatches</th>\n",
       "      <th>Gap Openings</th>\n",
       "      <th>Query Start</th>\n",
       "      <th>Query End</th>\n",
       "      <th>Target Start</th>\n",
       "      <th>Target End</th>\n",
       "      <th>E-Value</th>\n",
       "      <th>Bitscore</th>\n",
       "      <th>Length</th>\n",
       "      <th>C-terminus Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP003667212098</td>\n",
       "      <td>0.923</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>114</td>\n",
       "      <td>125</td>\n",
       "      <td>0.001832</td>\n",
       "      <td>33</td>\n",
       "      <td>128</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP001555959371</td>\n",
       "      <td>0.846</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>337</td>\n",
       "      <td>348</td>\n",
       "      <td>0.087220</td>\n",
       "      <td>28</td>\n",
       "      <td>352</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP003362106646</td>\n",
       "      <td>0.909</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>357</td>\n",
       "      <td>366</td>\n",
       "      <td>0.188500</td>\n",
       "      <td>27</td>\n",
       "      <td>367</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP001599236885</td>\n",
       "      <td>0.833</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>336</td>\n",
       "      <td>346</td>\n",
       "      <td>0.407200</td>\n",
       "      <td>26</td>\n",
       "      <td>351</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>all_cargo_clp_consensus</td>\n",
       "      <td>MGYP001627912779</td>\n",
       "      <td>0.875</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>120</td>\n",
       "      <td>127</td>\n",
       "      <td>1.290000</td>\n",
       "      <td>24</td>\n",
       "      <td>127</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Query            Target  Identity  Alignment Length  \\\n",
       "0  all_cargo_clp_consensus  MGYP003667212098     0.923                13   \n",
       "1  all_cargo_clp_consensus  MGYP001555959371     0.846                13   \n",
       "2  all_cargo_clp_consensus  MGYP003362106646     0.909                11   \n",
       "3  all_cargo_clp_consensus  MGYP001599236885     0.833                12   \n",
       "4  all_cargo_clp_consensus  MGYP001627912779     0.875                 8   \n",
       "\n",
       "   Mismatches  Gap Openings  Query Start  Query End  Target Start  Target End  \\\n",
       "0           0             1            1         13           114         125   \n",
       "1           1             1            1         13           337         348   \n",
       "2           0             1            1         11           357         366   \n",
       "3           1             1            1         12           336         346   \n",
       "4           1             0            6         13           120         127   \n",
       "\n",
       "    E-Value  Bitscore  Length  C-terminus Distance  \n",
       "0  0.001832        33     128                    3  \n",
       "1  0.087220        28     352                    4  \n",
       "2  0.188500        27     367                    1  \n",
       "3  0.407200        26     351                    5  \n",
       "4  1.290000        24     127                    0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "#First, let's make a dictionary where the keys are cargo MGYPs and values are the protein lengths\n",
    "length_dict = {str(record.id).split()[0]: len(str(record.seq)) for record in SeqIO.parse(\"../seqs/all_putative_cargo_proteins.fasta\", \"fasta\")}\n",
    "\n",
    "#Let's make a function to get the length of a cargo MGYP and apply it to the DataFrame\n",
    "def get_length(cargo_mgyp):\n",
    "    return(length_dict[cargo_mgyp])\n",
    "\n",
    "clp_df[\"Length\"] = clp_df[\"Target\"].apply(get_length)\n",
    "clp_df[\"C-terminus Distance\"] = clp_df[\"Length\"] - clp_df[\"Target End\"]\n",
    "clp_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's grab our Pfam data too:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The below code will load the Pfam labels dictionary downloaded from the GoogleResearch/Proteinfer GitHub repo\n",
    "# I've also manually added some code below to fill in some important missing labels that I manually curated\n",
    "\n",
    "import gzip\n",
    "import json\n",
    "\n",
    "with open(\"../DBs/label_descriptions.json.gz\", 'rb') as f:\n",
    "    with gzip.GzipFile(fileobj=f, mode='rb') as gzip_file:\n",
    "      labels_dict = json.load(gzip_file)\n",
    "\n",
    "labels_dict[\"PF19821\"] = \"Phage capsid protein\"\n",
    "labels_dict[\"PF19307\"] = \"Phage capsid-like protein\"\n",
    "labels_dict['PF19289'] = \"PmbA/TldA metallopeptidase C-terminal domain\"\n",
    "labels_dict['PF19290'] = \"PmbA/TldA metallopeptidase central domain\"\n",
    "labels_dict['PF20211'] = \"Family of unknown function (DUF6571)\"\n",
    "labels_dict['PF19782'] = \"Family of unknown function (DUF6267)\"\n",
    "labels_dict['PF19343'] = \"Family of unknown function (DUF5923)\"\n",
    "labels_dict['PF20036'] = \"Major capsid protein 13-like\"\n",
    "labels_dict['PF18960'] = \"Family of unknown function (DUF5702)\"\n",
    "labels_dict['PF18906'] = \"Phage tail tube protein\"\n",
    "labels_dict['PF19753'] = \"Family of unknown function (DUF6240)\"\n",
    "\n",
    "def get_label(pfam):\n",
    "    try:\n",
    "        return(labels_dict[pfam])\n",
    "    except KeyError:\n",
    "        return(None)\n",
    "\n",
    "pfam_df = pd.read_csv(\"../pfams/cargo_pfams.tsv\", sep=\"\\t\", names=[\"Cargo MGYP\", \"Pfam\", \"Start\", \"End\"])\n",
    "pfam_df[\"Description\"] = pfam_df[\"Pfam\"].apply(get_label)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now let's examine some of the CLP hits that are close to the N- or C-terminus:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Ferritin' 'Caspase recruitment domain']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Ferritin']\n",
      "['Ubiquinone biosynthesis protein COQ7']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Ferritin' 'Ferritin-like domain']\n",
      "['Rubrerythrin']\n",
      "['PH']\n",
      "['Death']\n",
      "['Death']\n",
      "['Chaperonin 10 Kd subunit' 'Chaperonin 10 Kd subunit']\n",
      "['Ferritin']\n",
      "['Death']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Ferritin' 'Death']\n",
      "['Death']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Death']\n",
      "['Dyp-type peroxidase family']\n",
      "['Dyp-type peroxidase family']\n",
      "['Caspase recruitment domain' 'Ferritin']\n",
      "['Ferritin']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Ferritin' 'Hemerythrin HHE cation binding domain']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Holin of 3TMs; for gene-transfer release']\n",
      "['Beta_propeller']\n",
      "['Beta_propeller']\n",
      "['Beta_propeller']\n",
      "['Primase C terminal 2 (PriCT-2)' 'VirE N-terminal domain']\n",
      "['Ribbon-helix-helix protein; copG family']\n"
     ]
    }
   ],
   "source": [
    "C_term_hits = clp_df[clp_df[\"C-terminus Distance\"] < 15].drop_duplicates(subset=\"Target\")\n",
    "\n",
    "for row in C_term_hits.to_dict(orient=\"records\"):\n",
    "    out_df = pfam_df[pfam_df[\"Cargo MGYP\"] == row[\"Target\"]][\"Description\"]\n",
    "    if len(out_df) > 0:\n",
    "        print(out_df.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Clp protease']\n",
      "['Clp protease']\n",
      "['Clp protease']\n",
      "['Clp protease']\n",
      "['Glycine rich protein']\n",
      "['Beta_propeller' 'Beta_propeller']\n",
      "[None]\n",
      "['Gluconate 2-dehydrogenase subunit 3']\n",
      "['Regulator of chromosome condensation (RCC1) repeat' 'Peptidase_CA']\n",
      "['Peptidase family M20/M25/M40' 'Peptidase dimerisation domain']\n",
      "['Peptidase dimerisation domain' 'Peptidase family M20/M25/M40']\n"
     ]
    }
   ],
   "source": [
    "N_term_hits = clp_df[clp_df[\"Target Start\"] < 15].drop_duplicates(subset=\"Target\")\n",
    "\n",
    "for row in N_term_hits.to_dict(orient=\"records\"):\n",
    "    out_df = pfam_df[pfam_df[\"Cargo MGYP\"] == row[\"Target\"]][\"Description\"]\n",
    "    if len(out_df) > 0:\n",
    "        print(out_df.values)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
